Set concurrent lumis/IOVs in ConfigBuilder if their value is different from default regardless of the number of threads

makortel opened the pull request:
#### PR description:

As discussed in Mattermost https://mattermost.web.cern.ch/cms-o-and-c/pl/sw55fqpr7pna8x581q453wcsne and reported in #37385, it can happen that the job configuration is generated with cmsDriver without specifying the number of threads (i.e. that is 1), and ends up being used with number of threads being overridden. If the number of concurrent lumis/IOVs is set different from the default in such case, that setting would be lost.

This PR overrides the `process.options.numberOfConcurrentLuminosityBlocks` and `process.options.eventSetup.numberOfConcurrentIOVs` when they have been changed from their defaults, regardless of the number of threads.

#### PR validation:

Checked with `limited,5.2,140.0,521.0,7.0,300.0,140.0,5.5,511.0,281.0,8.1,534.0,281.0,132.0,280.0,120.0` with `runTheMatrix.py` both `-t 2` option and without that the special cases that "need" to be explicitly set to use 1 concurrent lumi get that override in both cases. (I used the same set of workflows as in #35302)

makortel committed:
Set concurrent lumis/IOVs if their value is different from default regardless of the number of threads

It can happen that the job configuration is generated with cmsDriver
without specifying the number of threads (i.e. that is 1), and ends up
being used with number of threads being overridden. If the number of
concurrent lumis/IOVs is set different from the default in such case,
that setting would be lost.

makortel commented on Configuration/Applications/python/ConfigBuilder.py:
```diff
@@ -2268,19 +2268,31 @@ def prepare(self, doChecking = False):
+            # threads/streams can be set afterwards. The safest check
+            # is then to set assertLegacySafe == True only when
+            # explicitly asked to use 1 concurrent lumi.
+            self.pythonCfgCode +="if hasattr(process, 'DQMStore'): process.DQMStore.assertLegacySafe=cms.untracked.bool(False)\n"
```
@cms-sw/dqm-l2 I started to wonder if we could just set this parameter to `True` by default? (and eventually remove it as part of removal of support for legacy modules)

makortel replied:
Alternatively DQMStore could use `watchPreallocate` signal to check if the job has at least two concurrent lumis to disable this assert.

emanueleusai replied:
@makortel we support preventing concurrent lumis with the changes mentioned above until we complete the migration from legacy modules (and legacy way to access DQMStore). Either method is fine for us, thought it looks like the first one is simpler, so perhaps preferred.

makortel replied:
Sorry, I'm mixing things. The `assertLegacySafe` was set to `False` (which is what I should have written in the first message instead of `True`) by default already in https://github.com/cms-sw/cmssw/pull/34366. So this setting doesn't do anything anymore, and can be removed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37417/29093

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- Configuration/Applications (**operations**)


@cmsbuild, @perrotta, @qliphy, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @fabiocos this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test workflow 5.2,140.0,521.0,7.0,300.0,140.0,5.5,511.0,281.0,8.1,534.0,281.0,132.0,280.0,120.0

perrotta commented:
ping bot

smuzaffar commented:
let me see why bot did not run the cmparison job

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8b7d03/23547/summary.html
**COMMIT**: ee5fe8884b178dca0abdd3731c7de4cb5ed1a0e0
**CMSSW**: CMSSW_12_4_X_2022-03-30-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37417/23547/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/120.0_SinglePi0E10+SinglePi0E10+DIGI+RECO
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/132.0_LM9p+LM9p+DIGI+RECO
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/140.0_HydjetQ_B12_5020GeV_2011+HydjetQ_B12_5020GeV_2011+DIGIHI2011+RECOHI2011+HARVESTHI2011
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/280.0_AMPT_PPb_5020GeV_MinimumBias+AMPT_PPb_5020GeV_MinimumBias+DIGI+RECO+HARVEST
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/300.0_Pyquen_GammaJet_pt20_2760GeV+Pyquen_GammaJet_pt20_2760GeV+DIGIHIMIX+RECOHIMIX+HARVESTHI2018PPRECO
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/5.2_SingleMuPt10+SingleMuPt10FS+HARVESTFS
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/5.5_ZTT+ZTTFS+HARVESTFS
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/511.0_QCD_Pt-30_13TeV_aMCatNLO_herwig7+QCD_Pt-30_13TeV_aMCatNLO_herwig7+HARVESTGEN
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/521.0_WTolNuJets_LO_Mad_13TeV_py8_Ta+WTolNu01234Jets_5f_LO_MLM_Madgraph_LHE_13TeV+Hadronizer_TuneCP5_13TeV_MLM_5f_max4j_LHE_pythia8_Tauola+HARVESTGEN2
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/534.0_sherpa_ZtoLL_2j_MEPSatNLO_13TeV_MASTER+sherpa_ZtoLL_2j_MEPSatNLO_13TeV_MASTER+HARVESTGEN
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-8b7d03/8.1_BeamHalo_UP18+BeamHalo_UP18+DIGICOS_UP18+RECOCOS_UP18+ALCABH_UP18+HARVESTCOS_UP18

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 721 differences found in the comparisons
 - DQMHistoTests: Total files compared: 59
 - DQMHistoTests: Total histograms compared: 4058765
 - DQMHistoTests: Total failures: 3003
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 4055731
 - DQMHistoTests: Total skipped: 30
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 58 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 243 log files, 45 edm output root files, 59 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-30-1100+8b7d03/49335/triggerResults): found differences in 1 / 58 workflows

makortel committed:
Remove setting of DQMStore.assertLegacySafe as unnecessary

This parameter was set to False already in
https://github.com/cms-sw/cmssw/pull/34366.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37417/29131

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #37417 was updated. @cmsbuild, @perrotta, @qliphy, @fabiocos, @davidlange6 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8b7d03/23611/summary.html
**COMMIT**: 0f636f376dff6b70a40edc0abb0d9f4ae5969162
**CMSSW**: CMSSW_12_4_X_2022-04-01-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37417/23611/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3593039
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3593015
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-04-01-1100+8b7d03/49355/triggerResults): no differences found

jordan-martins commented:
Hi @qliphy @perrotta, 

I think is good to merge, right? We need this fix asap (and also backported to 12_2_x) in order to produce the Tau EGun samples needed for the EGM training. I know you just made a 12_2_3 but we will need another one (sorry!). Thanks, Jordan FYI @makortel

qliphy commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.

qliphy commented:
> Hi @qliphy @perrotta,
> 
> I think is good to merge, right? We need this fix asap (and also backported to 12_2_x) in order to produce the Tau EGun samples needed for the EGM training. I know you just made a 12_2_3 but we will need another one (sorry!). Thanks, Jordan FYI @makortel

@jordan-martins @makortel  A backport PR to 12_2_x/12_3_x is needed.

makortel commented:
Backports are here https://github.com/cms-sw/cmssw/pull/37445 and https://github.com/cms-sw/cmssw/pull/37446
